[Blueprint] Fix php warnings and improve error handling

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOPLUG-3933

Currently, the blueprint package has many php:ignore comments, incorrect type hints and no proper error handling.

This PR improves error handling, code quality, and documentation in the WooCommerce Blueprint package.


### Key Changes

1. **Error Handling**:
   - Added proper error handling for plugin deactivation
   - Enhanced file system operations with error checks
   - Improved error handling in theme/plugin resource storage

2. **Code Quality**:
   - Using WordPress filesystem functions instead of direct file operations
   - Added proper type hints and return type documentation
   - Fixed variable naming and added explanatory comments

3. **File System Operations**:
   - Added `wp_filesystem_get_contents` method
   - Improved file system initialization
   - Added path validation in JSON schema loading




<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Reviewing the code changes and ensure the changes make sense.

1. pnpm install
2. Enable the Blueprint feature `wp option set woocommerce_feature_blueprint_enabled yes`
3. Go to `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=blueprint`
4. Try to export and import a Blueprint. Confirm that everything works without errors.
5. Run `wp wc blueprint export test.json` and confirm that the blueprint export is successful.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Enhance UseWPFunctions trait with additional filesystem methods and improved documentation

- Added `wp_filesystem_get_contents` method for reading file contents.
- Updated return types in existing methods for clarity.
- Improved documentation comments for better understanding of method behaviors.

chihsuan committed:
Refactor ExportCli to utilize UseWPFunctions for filesystem operations

- Integrated UseWPFunctions trait into ExportCli class.
- Replaced direct file_put_contents call with wp_filesystem_put_contents for improved filesystem handling.
- Enhanced error handling for file saving operations.

chihsuan committed:
Refactor ResourceStorages class to improve parameter naming and enhance clarity

- Updated method parameters from `$resource` to `$resource_type` for better understanding.
- Adjusted logic in `is_supported_resource` and `download` methods to reflect the new parameter naming.
- Improved code readability by removing unnecessary comments and ensuring consistent formatting.

chihsuan committed:
Enhance OrgPluginResourceStorage class with improved error handling in download methods

- Updated return types in the `download` and `download_url` methods for clarity, changing from nullable types to more explicit error handling.
- Added checks for `WP_Error` to ensure robust error management during plugin downloads.
- Improved documentation comments to reflect the changes in return types and enhance understanding of method behaviors.

chihsuan committed:
Improve error handling in OrgThemeResourceStorage class by checking for WP_Error before accessing download link

- Added a check for WP_Error to return null if an error occurs, enhancing the robustness of the download link retrieval process.
- This change ensures that the method handles errors gracefully, preventing potential issues when accessing the download link.

chihsuan committed:
Update documentation for plugin activation and deletion methods in UsePluginHelpers trait

- Revised return type annotations for `activate_plugin_by_slug` and `delete_plugin_by_slug` methods to provide clearer expectations.
- Updated comments to reflect the possibility of returning `WP_Error` and `null`, enhancing understanding of error handling in these methods.

chihsuan committed:
Refactor plugin and theme activation processes for improved clarity

- Updated comments in `ImportActivatePlugin` to clarify the use of non-snake case property names.
- Modified `ImportActivateTheme` to remove the unnecessary return value from `wp_switch_theme`, enhancing code readability and understanding of the method's behavior.

chihsuan committed:
Refactor theme installation method for improved clarity

- Updated method parameter name from `$local_plugin_path` to `$local_path` to accurately reflect its purpose.
- Revised documentation comments to clarify that the method installs a theme from the local path, enhancing understanding of its functionality.

chihsuan committed:
Improve error handling in ImportDeactivatePlugin by checking for WP_Error after deactivating a plugin

- Added a check for WP_Error when attempting to deactivate a plugin, providing clearer feedback on success or failure.
- Updated the response messages to inform users whether the deactivation was successful or if an error occurred.

chihsuan committed:
Enhance JsonSchema class with improved error handling and filesystem integration

- Integrated UseWPFunctions trait to utilize WordPress filesystem functions.
- Added checks for valid schema path and improved error handling for JSON file reading.
- Updated validation to ensure 'steps' field is an array, enhancing schema integrity.

chihsuan committed:
Add changelog

chihsuan committed:
Refactor theme activation process to improve clarity and remove unnecessary return value

- Removed the return value from the `wp_switch_theme` method in `ImportActivateTheme`, enhancing code readability.
- Added a check to confirm the theme switch was successful before logging the debug message, improving the accuracy of feedback provided to users.

chihsuan committed:
Add tests for JsonSchema class to improve error handling coverage

- Introduced new test methods to validate exception handling for invalid paths and unreadable files.
- Utilized Mockery to simulate filesystem behavior, enhancing the robustness of the tests.
- Ensured that the JsonSchema class correctly throws exceptions for various error scenarios, improving overall error handling.

chihsuan committed:
Enhance ImportInstallPlugin with improved error handling and clarity

- Added checks to skip installation if the plugin is already installed and to validate resource types.
- Improved error handling during plugin installation and activation, providing clearer feedback on success or failure.
- Updated method documentation to reflect the possibility of returning WP_Error on failure.

louwie17 commented on packages/php/blueprint/src/Schemas/JsonSchema.php:
```diff
@@ -17,11 +21,24 @@ class JsonSchema {
+		$contents = $this->wp_filesystem_get_contents( $real_path );
+
+		if ( false === $contents ) {
+			throw new \RuntimeException( "Failed to read the JSON file at {$real_path}." );
```
If either of these exceptions get thrown the CLI would return a critical error, should we handle this more gracefully where `JsonSchema` is being used and let the user know the above errors? Similar to the warnings we show for the restricted options?
<img width="882" alt="Screenshot 2025-04-22 at 16 37 48" src="https://github.com/user-attachments/assets/837d44c7-d2a7-4306-9e6b-f5da88c20a1f" />

chihsuan replied:
Good point! I've updated the code to handle these exceptions more gracefully in 09ef550d20a0acf1ebb11c64e41bf03e764421bb

chihsuan replied:
<img width="1011" alt="Screenshot 2025-04-23 at 09 59 20" src="https://github.com/user-attachments/assets/bed4db31-539a-4691-881d-1c0c016516c4" />

louwie17 reviewed (commented):
Tested quite well, just left one comment around the JsonSchema testing and handling some of those errors better. Given on the front-end when manually uploading an invalid JSON, it does show an "Error reading or parsing file. Please check the schema." error.

chihsuan committed:
Implement error handling in ImportCli for schema file creation

- Added try-catch block in the run method to handle exceptions when creating the blueprint from the schema file.
- Improved user feedback by utilizing WP_CLI::error to display error messages, enhancing the robustness of the import process.

chihsuan commented:
> > Good point! I've updated the code to handle these exceptions more gracefully in [09ef550](https://github.com/woocommerce/woocommerce/commit/09ef550d20a0acf1ebb11c64e41bf03e764421bb)
> 
> @chihsuan it looks like this commit is not part of this PR but part of a different PR: #57384

Ah, sorry, @louwie17 I've pushed a commit to this branch: 4e223f52dff42abb2180f78a50fdef148498e8d4. I'll remove it from #57384.

kalessil committed:
Merge branch 'trunk' into wooplug-3933-fix-blueprint-type-warnings-and-improve-error-handling

louwie17 reviewed (approved):
Thanks for the updates @chihsuan this looks good and tested well! Nice work!
